pkg/beholder/client.go

```diff
@@ -0,0 +1,289 @@
+	if err != nil {
+		return noop, err
+	}
+	creds := insecure.NewCredentials()
+	if !cfg.InsecureConnection && cfg.CACertFile != "" {
+		creds, err = credentials.NewClientTLSFromFile(cfg.CACertFile, "")
+		if err != nil {
+			return noop, err
+		}
+	}
+	sharedLogExporter, err := otlploggrpcNew(
+		ctx,
+		otlploggrpc.WithTLSCredentials(creds),
+		otlploggrpc.WithEndpoint(cfg.OtelExporterGRPCEndpoint),
+	)
+	if err != nil {
+		return noop, err
+	}
+
+	// Logger
+	loggerProcessor := sdklog.NewBatchProcessor(
+		sharedLogExporter,
+		sdklog.WithExportTimeout(cfg.LogExportTimeout), // Default is 30s
+	)
+	loggerAttributes := []attribute.KeyValue{
+		attribute.String("beholder_data_type", "zap_log_message"),
+	}
+	loggerResource, err := sdkresource.Merge(
+		sdkresource.NewSchemaless(loggerAttributes...),
+		baseResource,
+	)
+	if err != nil {
+		return noop, err
+	}
+	loggerProvider := sdklog.NewLoggerProvider(
+		sdklog.WithResource(loggerResource),
+		sdklog.WithProcessor(loggerProcessor),
+	)
+	logger := loggerProvider.Logger(cfg.PackageName)
+
+	// Tracer
+	tracerProvider, err := newTracerProvider(cfg, baseResource, creds)
+	if err != nil {
+		return noop, err
+	}
+	tracer := tracerProvider.Tracer(cfg.PackageName)
+
+	// Meter
+	meterProvider, err := newMeterProvider(cfg, baseResource, creds)
+	if err != nil {
+		return noop, err
+	}
+	meter := meterProvider.Meter(cfg.PackageName)
+
+	// Message Emitter
+	messageLogProcessor := sdklog.NewBatchProcessor(
+		sharedLogExporter,
+		sdklog.WithExportTimeout(cfg.EmitterExportTimeout), // Default is 30s
+	)
+	messageAttributes := []attribute.KeyValue{
+		attribute.String("beholder_data_type", "custom_message"),
+	}
+	messageLoggerResource, err := sdkresource.Merge(
+		sdkresource.NewSchemaless(messageAttributes...),
+		baseResource,
+	)
+	if err != nil {
+		return noop, err
+	}
+	messageLoggerProvider := sdklog.NewLoggerProvider(
+		sdklog.WithResource(messageLoggerResource),
+		sdklog.WithProcessor(messageLogProcessor),
+	)
+	messageLogger := messageLoggerProvider.Logger(cfg.PackageName)
+	messageEmitter := newMessageEmitter(sharedLogExporter, messageLogger)
+
+	setOtelErrorHandler(errorHandler)
+
+	client := OtelClient{cfg, logger, tracer, meter, messageEmitter, loggerProvider, tracerProvider, meterProvider}
+
+	return client, nil
+}
+
+func (c OtelClient) SetGlobals() {
+	// Logger
+	otelglobal.SetLoggerProvider(c.LoggerProvider)
+	// Tracer
+	otel.SetTracerProvider(c.TracerProvider)
+	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{}))
+	// Meter
+	otel.SetMeterProvider(c.MeterProvider)
+}
+
+func (c OtelClient) Close() (err error) {
+	for _, provider := range []any{c.LoggerProvider, c.TracerProvider, c.MeterProvider} {
+		if p, ok := provider.(otelProvider); ok {
+			err = errors.Join(err, p.Shutdown(context.Background()))
+		} else {
+			err = errors.Join(err, errors.New("provider does not implement Shutdown method"))
```

Comment:
can we do this at build time rather than run time?

Reply:
@patrickhuie19 https://github.com/smartcontractkit/chainlink-common/pull/694/commits/b80e230cd57a43ccf512f67c615aa5913148270b